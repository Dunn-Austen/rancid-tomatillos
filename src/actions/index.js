export const addMovies = movies => ({
  type: 'ADD_MOVIES',
  movies: movies
});

export const updateUser = user => ({
  type: 'UPDATE_USER',
  user: user
});

export const updateLoggedInStatus = status => ({
  type: 'UPDATE_LOGGEDIN_STATUS',
  status: status
});

export const handleError = errorMessage => ({
  type: 'HANDLE_ERROR',
  errorMessage
});

export const isLoading = loadingStatus => ({
  //Passing in a boolean, where loadingStatus key is assigned: true/false
  type: 'IS_LOADING',
  loadingStatus
});
