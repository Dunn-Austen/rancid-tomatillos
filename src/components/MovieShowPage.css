.movie-page {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.backdrop {
  height: 400px;
  width: 400px;
}

.overview {
  text-align: center;
}
