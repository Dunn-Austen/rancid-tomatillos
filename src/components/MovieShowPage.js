import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import './MovieShowPage.css';
import { updateUser, handleError } from '../actions/index';
import { updateRatings, fetchRatings, deleteRating } from '../apiCalls';
import { findRating } from '../util';

export class MovieShowPage extends Component {
  // constructor(props) {
  //   super(props);
  // }

  handleRatingsUpdates = event => {
    const { user, id, updateUser } = this.props;
    updateRatings(id, Number(event.target.value), user.id)
      .then(data => {
        fetchRatings(user.id).then(ratingData => {
          const newRatings = { ...user, ratings: ratingData.ratings };
          updateUser(newRatings);
        });
      })
      .catch(error => {
        this.props.handleError(
          'Problem updating your rating, please try again later.'
        );
      });
  };

  render() {
    const {
      id,
      title,
      backdrop_path,
      release_date,
      overview,
      average_rating,
      isLoggedIn,
      updateUser,
      user,
      errorMessage
    } = this.props;

    return (
      <div className='movie-page'>
        <img className='backdrop' src={backdrop_path} alt={title} />
        <h1 className='movie__title--two'>{title}</h1>
        <p className='in-theaters'>In Theaters: {release_date}</p>
        <p className='overview'>{overview}</p>
        <p className='average__rating--two'>
          Average Rating: {Math.round(average_rating * 10) / 10}
        </p>
        {isLoggedIn && findRating(id, user, 'rating') !== '...' && (
          <div className='already-rated'>
            <p
              className={
                findRating(id, user, 'rating') === 10
                  ? 'top-marks'
                  : 'my-rating'
              }>
              My Rating: {findRating(id, user, 'rating')}
            </p>
            <button
              className='reset-rating'
              onClick={() => {
                deleteRating(findRating(id, user, 'id'), user.id)
                  .then(data => {
                    fetchRatings(user.id).then(ratingData => {
                      const updatedRatings = {
                        ...user,
                        ratings: ratingData.ratings
                      };
                      updateUser(updatedRatings);
                    });
                  })
                  .catch(error => {
                    this.props.handleError(
                      'Problem deleting your rating, please try again later.'
                    );
                  });
              }}>
              Reset Rating
            </button>
          </div>
        )}
        {isLoggedIn && findRating(id, user, 'rating') === '...' && (
          <div className='rating-bar'>
            <div className='rating-scale'>
              {errorMessage && (
                <h1 className='error-styling'>{this.props.errorMessage}</h1>
              )}
              <button
                className='rating-btn btn-one'
                value='1'
                onClick={event => this.handleRatingsUpdates(event)}>
                1
              </button>
              <button
                className='rating-btn btn-two'
                value='2'
                onClick={event => this.handleRatingsUpdates(event)}>
                2
              </button>
              <button
                className='rating-btn btn-three'
                value='3'
                onClick={event => this.handleRatingsUpdates(event)}>
                3
              </button>
              <button
                className='rating-btn btn-four'
                value='4'
                onClick={event => this.handleRatingsUpdates(event)}>
                4
              </button>
              <button
                className='rating-btn btn-five'
                value='5'
                onClick={event => this.handleRatingsUpdates(event)}>
                5
              </button>
              <button
                className='rating-btn btn-six'
                value='6'
                onClick={event => this.handleRatingsUpdates(event)}>
                6
              </button>
              <button
                className='rating-btn btn-seven'
                value='7'
                onClick={event => this.handleRatingsUpdates(event)}>
                7
              </button>
              <button
                className='rating-btn btn-eight'
                value='8'
                onClick={event => this.handleRatingsUpdates(event)}>
                8
              </button>
              <button
                className='rating-btn btn-nine'
                value='9'
                onClick={event => this.handleRatingsUpdates(event)}>
                9
              </button>
              <button
                className='rating-btn btn-ten'
                value='10'
                onClick={event => this.handleRatingsUpdates(event)}>
                10
              </button>
            </div>
          </div>
        )}
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  isLoggedIn: state.isLoggedIn,
  user: state.user,
  errorMessage: state.errorMessage
});

export const mapDispatchToProps = dispatch => ({
  updateUser: user => dispatch(updateUser(user)),
  handleError: errorMessage => dispatch(handleError(errorMessage))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieShowPage);

MovieShowPage.propTypes = {
  allMovies: PropTypes.array,
  isLoggedIn: PropTypes.bool,
  user: PropTypes.object,
  updateUser: PropTypes.func,
  errorMessage: PropTypes.string,
  handleError: PropTypes.func
};
