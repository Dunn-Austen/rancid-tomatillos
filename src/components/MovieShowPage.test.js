import React from 'react';
import { shallow } from 'enzyme';
import { MovieShowPage, mapStateToProps, mapDispatchToProps } from './MovieShowPage';
import { updateUser} from '../actions'

describe('MovieShowPage', () => {
  let wrapper, mockUpdateUser;

  beforeEach(() => {
    mockUpdateUser = jest.fn();
    wrapper = shallow(<MovieShowPage
      id={2}
      title="Ad Astra"
      poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
      release_date="2019-09-17"
      overview="The near future, humanity to look to the stars an....."
      average_rating={5.142857142857143}
      isLoggedIn={true}
      updateUser={mockUpdateUser}
      user={ {id: 22, name: 'Marge', email: 'marge@turing.io', ratings: [
        {id:45,
          user_id:9,
          movie_id:8,
          rating:8,
          created_at:"2019-12-25T20:16:34.893Z",
          updated_at:"2019-12-25T20:16:34.893Z"},
        {id:46,
          user_id:9,
          movie_id:10,
          rating:10,
          created_at:"2019-12-25T20:30:21.606Z",
          updated_at:"2019-12-25T20:30:21.606Z"}
        ]} }
    />)
  })

  it('should match the MovieShowPage Snapshot', () => {
    wrapper = shallow(<MovieShowPage
      id={2}
      title="Ad Astra"
      poster_path="https://image.tmdb.org/t/p/original//xBHvZcjRiWyobQ9kxBhO6B2dtRI.jpg"
      backdrop_path="https://image.tmdb.org/t/p/original//5BwqwxMEjeFtdknRV792Svo0K1v.jpg"
      release_date="2019-09-17"
      overview="The near future, humanity to look to the stars an....."
      average_rating={5.142857142857143}
      isLoggedIn={false}
      updateUser={mockUpdateUser}
      user={ {} }
      />);
    expect(wrapper).toMatchSnapshot();
  });

// Should have snapshop test for each conditional rendering possibility
// Do we need to pass the props through for proper snapshot?
// 'should match the MovieShowPage snapshot when no user logged in'
// 'should match snapshot when logged in and user rating < 10'
// 'should match snapshot when logged in and user rating === 10'
// 'should match snapshot when logged in and no user rating yet'

  //
  // it('should', () => {
  //
  // });
  //
  // it('should', () => {
  //
  // });
  //
  // it('should', () => {
  //
  // });
  //
  // it('should', () => {
  //
  // });




  describe('mapsStateToProps', () => {
    it('should return only the necessary information from the redux store', () => {
      const mockState = {
        movies: [],
        user: {newUser: 'Tron'},
        isLoggedIn: true,
        errorMessage: '',
        loadingStatus: false,
      };
      const expected = {
        isLoggedIn: true,
        user: {newUser: 'Tron'}
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('calls dispatch with an updateUser action when updateUser is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = updateUser({some: 'thing'});
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.updateUser({some: 'thing'});

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
