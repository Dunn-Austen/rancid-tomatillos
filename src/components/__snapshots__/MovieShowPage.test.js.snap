// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieShowPage should match the MovieShowPage Snapshot 1`] = `
<div
  className="movie-page"
>
  <img
    className="backdrop"
  />
  <h1 />
  <p>
    In Theaters: 
  </p>
  <p
    className="overview"
  />
  <p>
    Average Rating: 
  </p>
  <div
    className="rating-bar"
  >
    <p
      className="rating-description"
    >
      Rate Film:
    </p>
    <div
      className="rating-scale"
    >
      <button
        className="rating-btn btn-one"
        value="1"
      >
        1
      </button>
      <button
        className="rating-btn btn-two"
        value="2"
      >
        2
      </button>
      <button
        className="rating-btn btn-three"
        value="3"
      >
        3
      </button>
      <button
        className="rating-btn btn-four"
        value="4"
      >
        4
      </button>
      <button
        className="rating-btn btn-five"
        value="5"
      >
        5
      </button>
      <button
        className="rating-btn btn-six"
        value="6"
      >
        6
      </button>
      <button
        className="rating-btn btn-seven"
        value="7"
      >
        7
      </button>
      <button
        className="rating-btn btn-eight"
        value="8"
      >
        8
      </button>
      <button
        className="rating-btn btn-nine"
        value="9"
      >
        9
      </button>
      <button
        className="rating-btn btn-ten"
        onClick={[Function]}
        value="10"
      >
        10
      </button>
    </div>
  </div>
</div>
`;
